File: AIQuotaBar/Tests/ModelUtilizationHistoryStoreTests.swift
import XCTest
@testable import AIQuotaBar

/// 覆盖 ModelUtilizationHistoryStore 的 load/save 路径:
/// - 损坏文件 (JSON 解析失败 / schema version 不匹配) → move 到 corrupt/ 备份 + 返回空
/// - 读失败 → **不应** move 文件（防止数据锁死在备份里）
/// - 正常 save/load 等幂
/// - save 失败 → 发 didFailToSaveHistory 通知
final class ModelUtilizationHistoryStoreTests: XCTestCase {
    private var tempDir: URL!
    private var store: ModelUtilizationHistoryStore!

    override func setUp() {
        super.setUp()
        tempDir = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("util-test-\(UUID().uuidString)", isDirectory: true)
        try? FileManager.default.createDirectory(at: tempDir, withIntermediateDirectories: true)
        store = ModelUtilizationHistoryStore(directoryURL: tempDir)
    }

    override func tearDown() {
        try? FileManager.default.removeItem(at: tempDir)
        super.tearDown()
    }

    // MARK: - Load 路径

    func test_noFile_returnsEmptyWithoutCorruptDir() {
        let result = store.load(for: .miniMax)
        XCTAssertTrue(result.historiesOrEmpty.isEmpty, "不存在的文件应返回空 store")

        // 没有文件 → 不应触发 backup,corrupt/ 目录也不应被创建
        let corruptDir = tempDir.appendingPathComponent("corrupt")
        XCTAssertFalse(FileManager.default.fileExists(atPath: corruptDir.path),
                       "无文件场景下不应创建 corrupt/ 目录")
    }

    func test_corruptJSON_movesToCorruptBackup() {
        let provider = UsageProvider.glm
        let url = store.fileURL(for: provider)
        try? "{ invalid json".data(using: .utf8)?.write(to: url, options: [.atomic])

        let result = store.load(for: provider)

        XCTAssertTrue(result.historiesOrEmpty.isEmpty, "损坏 JSON 应返回空 store")
        XCTAssertFalse(FileManager.default.fileExists(atPath: url.path),
                       "损坏文件应被 move 到 corrupt/,原路径应不存在")
        XCTAssertTrue(FileManager.default.fileExists(atPath: corruptDirPath()),
                      "corrupt/ 目录应被创建")
    }

    func test_schemaVersionMismatch_movesToCorruptBackup() {
        let provider = UsageProvider.miniMax
        let url = store.fileURL(for: provider)
        try? #"{ "version": 999, "histories": {} }"#.data(using: .utf8)?.write(to: url, options: [.atomic])

        let result = store.load(for: provider)

        XCTAssertTrue(result.historiesOrEmpty.isEmpty, "新版本 schema 应返回空 store")
        XCTAssertFalse(FileManager.default.fileExists(atPath: url.path),
                       "version 不匹配文件应被 move 到 corrupt/")
    }

    func test_corruptFile_postsBackupNotification() {
        let provider = UsageProvider.codex
        let url = store.fileURL(for: provider)
        try? "not json".data(using: .utf8)?.write(to: url, options: [.atomic])

        let expectation = expectation(forNotification: ModelUtilizationHistoryStore.didBackupCorruptHistory,
                                      object: nil)

        _ = store.load(for: provider)

        wait(for: [expectation], timeout: 1.0)
    }

    // MARK: - Round-trip

    func test_saveLoad_roundTrip_singleModel() {
        let provider = UsageProvider.miniMax
        let history = ModelUtilizationHistory(
            modelId: "test:single",
            entries: [
                UtilizationHistoryEntry(
                    capturedAt: Date(timeIntervalSince1970: 1_700_000_000),
                    usedPercent: 50,
                    resetsAt: Date(timeIntervalSince1970: 1_700_500_000)
                )
            ]
        )
        let payload = ModelUtilizationStoreData(histories: ["test:single": history])

        store.save(payload, for: provider)
        let loaded = store.load(for: provider)

        XCTAssertEqual(loaded.historiesOrEmpty["test:single"]?.entries.count, 1)
        XCTAssertEqual(loaded.historiesOrEmpty["test:single"]?.entries.first?.usedPercent, 50)
    }

    func test_saveLoad_roundTrip_multipleModels() {
        let provider = UsageProvider.codex
        let modelA = ModelUtilizationHistory(
            modelId: "codex:a",
            entries: [
                UtilizationHistoryEntry(
                    capturedAt: Date(timeIntervalSince1970: 1_700_000_000),
                    usedPercent: 30,
                    resetsAt: Date(timeIntervalSince1970: 1_700_500_000)
                )
            ]
        )
        let modelB = ModelUtilizationHistory(
            modelId: "codex:b",
            entries: [
                UtilizationHistoryEntry(
                    capturedAt: Date(timeIntervalSince1970: 1_700_000_100),
                    usedPercent: 70,
                    resetsAt: nil
                )
            ]
        )
        let payload = ModelUtilizationStoreData(histories: ["codex:a": modelA, "codex:b": modelB])

        store.save(payload, for: provider)
        let loaded = store.load(for: provider)

        XCTAssertEqual(loaded.historiesOrEmpty["codex:a"]?.entries.first?.usedPercent, 30)
        XCTAssertEqual(loaded.historiesOrEmpty["codex:b"]?.entries.first?.usedPercent, 70)
        XCTAssertNil(loaded.historiesOrEmpty["codex:b"]?.entries.first?.resetsAt)
    }

    // MARK: - Save failure

    func test_saveFailure_postsNotification() {
        // 构造一个 directoryURL 让 createDirectory 失败:把 directoryURL 指向一个已存在的普通文件。
        let blockingFile = tempDir.appendingPathComponent("blocking-file")
        try? "x".write(to: blockingFile, atomically: true, encoding: .utf8)

        let brokenStore = ModelUtilizationHistoryStore(directoryURL: blockingFile)
        let payload = ModelUtilizationStoreData(histories: ["k": ModelUtilizationHistory(modelId: "k")])

        let expectation = expectation(forNotification: ModelUtilizationHistoryStore.didFailToSaveHistory,
                                      object: nil)

        brokenStore.save(payload, for: .glm)

        wait(for: [expectation], timeout: 1.0)
    }

    // MARK: - Helpers

    private func corruptDirPath() -> String {
        tempDir.appendingPathComponent("corrupt").path
    }
}

File: windows/tests/AIQuotaBar.Core.Tests/Quota/ModelUtilizationHistoryStoreTests.cs
// Swift 来源：AIQuotaBar/Tests/ModelUtilizationHistoryStoreTests.swift — final class
// ModelUtilizationHistoryStoreTests（被测：AIQuotaBar.Core/Quota/ModelUtilizationHistoryStore.cs）。
//
// 验收对照表（Swift 函数 → xUnit 方法）：
// - test_noFile_returnsEmptyWithoutCorruptDir       → NoFile_ReturnsEmptyWithoutCorruptDir
// - test_corruptJSON_movesToCorruptBackup           → CorruptJson_MovesToCorruptBackup
// - test_schemaVersionMismatch_movesToCorruptBackup → SchemaVersionMismatch_MovesToCorruptBackup
// - test_corruptFile_postsBackupNotification        → CorruptFile_PostsBackupNotification
// - test_saveLoad_roundTrip_singleModel             → SaveLoad_RoundTrip_SingleModel
// - test_saveLoad_roundTrip_multipleModels          → SaveLoad_RoundTrip_MultipleModels
// - test_saveFailure_postsNotification              → SaveFailure_PostsNotification
//
// 差异说明：Swift NotificationCenter 通知（didBackupCorruptHistory / didFailToSaveHistory）
// 在 C# 端为 store 实例事件；expectation(forNotification:) 等价改为事件标志位断言
// （LoadAsync/SaveAsync 返回前同步触发事件，无需异步等待）。
// store 文件是本地持久化格式：损坏样本（"{ invalid json" / "not json" / version 999）与
// Swift 端逐字相同，不属于 API 响应 fixtures 范畴。

#nullable enable

using System;
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;

using AIQuotaBar.Core.Contracts;
using AIQuotaBar.Core.Quota;

using Xunit;

namespace AIQuotaBar.Core.Tests.Quota;

/// <summary>
/// 覆盖 ModelUtilizationHistoryStore 的 load/save 路径（Swift 端类注释逐条对应）：
/// 损坏文件（JSON 解析失败 / schema version 不匹配）→ move 到 corrupt/ 备份 + 返回空；
/// 读失败 → 不应 move 文件（防止数据锁死在备份里）；正常 save/load 等幂；
/// save 失败 → DidFailToSaveHistory 事件。
/// </summary>
public sealed class ModelUtilizationHistoryStoreTests : IDisposable
{
    private readonly string _tempDir;
    private readonly ModelUtilizationHistoryStore _store;

    public ModelUtilizationHistoryStoreTests()
    {
        _tempDir = Path.Combine(Path.GetTempPath(), "util-test-" + Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(_tempDir);
        _store = new ModelUtilizationHistoryStore(_tempDir);
    }

    public void Dispose()
    {
        if (Directory.Exists(_tempDir))
        {
            Directory.Delete(_tempDir, recursive: true);
        }
    }

    // ---------------------------------------------------------------- load 路径

    [Fact]
    public async Task NoFile_ReturnsEmptyWithoutCorruptDir()
    {
        var result = await _store.LoadAsync(UsageProvider.MiniMax);
        Assert.True(result.HistoriesOrEmpty().Count == 0, "不存在的文件应返回空 store");

        // 没有文件 → 不应触发 backup，corrupt/ 目录也不应被创建。
        var corruptDir = CorruptDirPath();
        Assert.False(Directory.Exists(corruptDir), "无文件场景下不应创建 corrupt/ 目录");
    }

    [Fact]
    public async Task CorruptJson_MovesToCorruptBackup()
    {
        var provider = UsageProvider.Glm;
        var url = _store.FileFor(provider);
        await File.WriteAllTextAsync(url, "{ invalid json");

        var result = await _store.LoadAsync(provider);

        Assert.True(result.HistoriesOrEmpty().Count == 0, "损坏 JSON 应返回空 store");
        Assert.False(File.Exists(url), "损坏文件应被 move 到 corrupt/，原路径应不存在");
        Assert.True(Directory.Exists(CorruptDirPath()), "corrupt/ 目录应被创建");
    }

    [Fact]
    public async Task SchemaVersionMismatch_MovesToCorruptBackup()
    {
        var provider = UsageProvider.MiniMax;
        var url = _store.FileFor(provider);
        await File.WriteAllTextAsync(url, "{ \"version\": 999, \"histories\": {} }");

        var result = await _store.LoadAsync(provider);

        Assert.True(result.HistoriesOrEmpty().Count == 0, "新版本 schema 应返回空 store");
        Assert.False(File.Exists(url), "version 不匹配文件应被 move 到 corrupt/");
    }

    [Fact]
    public async Task CorruptFile_PostsBackupNotification()
    {
        var provider = UsageProvider.Codex;
        var url = _store.FileFor(provider);
        await File.WriteAllTextAsync(url, "not json");

        var didBackup = false;
        _store.DidBackupCorruptHistory += (_, _) => didBackup = true;

        await _store.LoadAsync(provider);

        Assert.True(didBackup, "损坏文件被备份后应触发 DidBackupCorruptHistory 事件");
    }

    // ---------------------------------------------------------------- round-trip

    [Fact]
    public async Task SaveLoad_RoundTrip_SingleModel()
    {
        var provider = UsageProvider.MiniMax;
        var history = new ModelUtilizationHistory(
            ModelId: "test:single",
            Entries: new[]
            {
                new UtilizationHistoryEntry(
                    CapturedAt: new DateTimeOffset(1_700_000_000, TimeSpan.Zero),
                    UsedPercent: 50,
                    ResetsAt: new DateTimeOffset(1_700_500_000, TimeSpan.Zero)),
            });
        var payload = new ModelUtilizationStoreData(
            Histories: new Dictionary<string, ModelUtilizationHistory> { ["test:single"] = history });

        await _store.SaveAsync(payload, provider);
        var loaded = await _store.LoadAsync(provider);

        Assert.True(
            loaded.HistoriesOrEmpty().TryGetValue("test:single", out var restored),
            "round-trip 后应存在 key test:single");
        Assert.Equal(1, restored!.Entries.Count);
        Assert.Equal(50, restored.Entries[0].UsedPercent);
    }

    [Fact]
    public async Task SaveLoad_RoundTrip_MultipleModels()
    {
        var provider = UsageProvider.Codex;
        var modelA = new ModelUtilizationHistory(
            ModelId: "codex:a",
            Entries: new[]
            {
                new UtilizationHistoryEntry(
                    CapturedAt: new DateTimeOffset(1_700_000_000, TimeSpan.Zero),
                    UsedPercent: 30,
                    ResetsAt: new DateTimeOffset(1_700_500_000, TimeSpan.Zero)),
            });
        var modelB = new ModelUtilizationHistory(
            ModelId: "codex:b",
            Entries: new[]
            {
                new UtilizationHistoryEntry(
                    CapturedAt: new DateTimeOffset(1_700_000_100, TimeSpan.Zero),
                    UsedPercent: 70,
                    ResetsAt: null),
            });
        var payload = new ModelUtilizationStoreData(
            Histories: new Dictionary<string, ModelUtilizationHistory>
            {
                ["codex:a"] = modelA,
                ["codex:b"] = modelB,
            });

        await _store.SaveAsync(payload, provider);
        var loaded = await _store.LoadAsync(provider);

        Assert.True(
            loaded.HistoriesOrEmpty().TryGetValue("codex:a", out var restoredA),
            "round-trip 后应存在 key codex:a");
        Assert.Equal(30, restoredA!.Entries[0].UsedPercent);
        Assert.True(
            loaded.HistoriesOrEmpty().TryGetValue("codex:b", out var restoredB),
            "round-trip 后应存在 key codex:b");
        Assert.Equal(70, restoredB!.Entries[0].UsedPercent);
        Assert.Null(restoredB!.Entries[0].ResetsAt);
    }

    // ---------------------------------------------------------------- save 失败

    [Fact]
    public async Task SaveFailure_PostsNotification()
    {
        // 构造一个 directoryPath 让 CreateDirectory 失败：把它指向一个已存在的普通文件。
        var blockingFile = Path.Combine(_tempDir, "blocking-file");
        await File.WriteAllTextAsync(blockingFile, "x");

        var brokenStore = new ModelUtilizationHistoryStore(blockingFile);
        var payload = new ModelUtilizationStoreData(
            Histories: new Dictionary<string, ModelUtilizationHistory>
            {
                ["k"] = new ModelUtilizationHistory(ModelId: "k", Entries: Array.Empty<UtilizationHistoryEntry>()),
            });

        var didFail = false;
        brokenStore.DidFailToSaveHistory += (_, _) => didFail = true;

        await brokenStore.SaveAsync(payload, UsageProvider.Glm);

        Assert.True(didFail, "保存失败应触发 DidFailToSaveHistory 事件");
    }

    // ---------------------------------------------------------------- helpers

    // Swift: private func corruptDirPath() -> String。
    private string CorruptDirPath() => Path.Combine(_tempDir, ModelUtilizationHistoryStore.CorruptBackupDirectoryName);
}
